atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadSentinelHandler.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl;
+
+import com.google.common.collect.Maps;
+import com.palantir.atlasdb.cell.api.TransactionKeyValueService;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.transaction.api.Transaction.TransactionType;
+import com.palantir.atlasdb.transaction.api.TransactionFailedRetriableException;
+import com.palantir.atlasdb.transaction.api.TransactionReadSentinelBehavior;
+import com.palantir.atlasdb.transaction.service.TransactionService;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public final class ReadSentinelHandler {
+    private final TransactionKeyValueService transactionKeyValueService;
+    private final TransactionService transactionService;
+    private final Set<Cell> trackedOrphanedSentinels;
+    private final TransactionReadSentinelBehavior readSentinelBehavior;
+    private final OrphanedSentinelDeleter orphanedSentinelDeleter;
+
+    public ReadSentinelHandler(
```

Comment:
In general goes without saying that we can probs now write some smaller tests for these things? I don't know how well this is covered in the big tests

Reply:
This one specifically is covered well (I wrote it, see #5231) 😅. Yeah we can do that, though that's more for the productionisation stage